### PR TITLE
feat(middleware): Add middleware method "process_resource"

### PR DESCRIPTION
**Pull request by rsyed83**
Create a new middleware method process_resource that occurs after process_request and routing, and ensure that the resource object is accessible. Additionally, make the resource object available to process_response.

Issue #400

### CODE REVIEW

File: falcon/api.py [hunk 1]
```diff
@@ -52,66 +52,27 @@ def process_request(self, req, resp):
                                 the on_* responder
                         \"""
 
-                    def process_response(self, req, resp)
-                        \"""Post-processing of the response (after routing).
+                    def process_resource(self, req, resp, resource):
+                        \"""Process the request after routing.
+
+                        Args:
+                            req: Request object that will eventually be
```
Comment (kgriffs) [inline falcon/api.py]: At this point, the routing has occurred, so maybe reword this to:

"req: Request object that will be passed to the routed responder"

File: falcon/api.py [hunk 2]
```diff
@@ -52,66 +52,27 @@ def process_request(self, req, resp):
...
+                                routed to an on_* responder method
+                            resp: Response object that will be routed to
```
Comment (kgriffs) [inline falcon/api.py]: Similar comment to above. Suggested text:

"resp: Response object that will be passed to the responder"

File: falcon/api.py [hunk 3]
```diff
@@ -52,66 +52,27 @@ def process_request(self, req, resp):
...
-            Finally, if one of the *process_response* methods raises an error,
-            or the routed on_* responder method itself raises an error, the
-            exception will be handled in a similar manner as above. Then,
-            the framework will execute any remaining middleware on the
-            stack.
+                    def process_response(self, req, resp, resource)
+                        \"""Post-processing of the response (after routing).
 
+                        Args:
+                            req: Request object that will eventually be
```
Comment (kgriffs) [inline falcon/api.py]: At this point, the request was routed to a responder, if one could be found. But even if no route is found, the middleware component's `process_response` method will still be called with the usual Request and Response objects. Rather than trying to explain all this, it may be best to keep things simple in the docstring, e.g.:

```
req: Request object
resp: Response object
```

File: falcon/api.py [hunk 4]
```diff
@@ -52,66 +52,27 @@ def process_request(self, req, resp):
...
+                                routed to an on_* responder method
+                            resp: Response object that will be routed to
+                                the on_* responder
+                            resource: Resource object that owns the responder
```
Comment (kgriffs) [inline falcon/api.py]: I think it would be helpful to mention that this argument may be `None`:

"resource: Resource object to which the request was routed. May be *None* if no route was found for the request."

File: tests/test_middlewares.py [hunk 5]
```diff
@@ -299,6 +321,38 @@ def handler(ex, req, resp, params):
...
+
+            def process_resource(self, req, resp, resource):
+                raise Exception("Always fail")
+
+        self.api = falcon.API(middleware=[ExecutedFirstMiddleware(),
+                                          RaiseErrorMiddleware(),
+                                          ExecutedLastMiddleware()])
+
+        def handler(ex, req, resp, params):
+            context['error_handler'] = True
```
Comment (kgriffs) [inline tests/test_middlewares.py]: Missing assertion? `context['error_handler']` is set here, but I can't find anywhere the value is being checked...

File: doc/api/middleware.rst [hunk 6]
```diff
@@ -0,0 +1,108 @@
+.. _middleware:
+
+Middleware
+==========
+
+Middleware components execute both before and after the framework
+routes the request.  Middleware registration is achieved through
```
Comment (kgriffs) [inline doc/api/middleware.rst]: I think we can make the second sentence flow a little better, e.g.:

```rst
Middleware is registered by passing components to the :ref:`API class <api>` initializer.
```

File: doc/api/middleware.rst [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+            Args:
+                req: Request object that will eventually be
+                    routed to an on_* responder method
+                resp: Response object that will be routed to
+                    the on_* responder
+                resource: Resource object that owns the responder
+            """
+
+
+Because middleware can execute before routing as occurred, if a
```
Comment (kgriffs) [inline doc/api/middleware.rst]: typo: "as" ==> "has"

**Comment by ealogar** [comment]
Nice work!

**Comment by kgriffs** [comment]
Nice job, LGTM.

**Comment by rsyed83** [comment]
Thanks guys!  My first open source contribution ever (long overdue).
